book/src/porting/implementing_sysdeps.md

```diff
@@ -0,0 +1,153 @@
+
+For the demo OS, we'll symlink each ABI header to Linux's definition in `abis/linux`. Then, we'll add the following to our `meson.build`:
+
+```meson
+{{#include ../../../sysdeps/demo/meson.build:abi-includes}}
+```
+
+This is not an exhaustive list of ABI headers but should be enough to get started.
+
+## Implementing (S)crt1
+
+Next, we must provide a definition of the ELF entry point (traditionally named `_start`). Each ISA tends to require its own definition written in assembly; for example, RISC-V targets must initialise the `gp` register before jumping to C++.
+
+Traditionally the file that defines `_start` is called `crt1.S` (or `Scrt1.S` for position independent executables). This produces an object file which has to be linked into every application.
+
+Part of configuring an [OS Specific Toolchain](https://wiki.osdev.org/OS_Specific_Toolchain) is [specifying the location of `(S)crt1.o`](https://wiki.osdev.org/OS_Specific_Toolchain#Start_Files_Directory) so that it is linked automatically. Generic ELF targets must link it explicitly, as the compiler does not know about it.
+
+We recommend copying `(S)crt1.S` from an existing target like Linux. Then, we'll compile it by adding the following to our `meson.build`:
+
+```
+{{#include ../../../sysdeps/demo/meson.build:crt1}}
+```
+
+## Implementing a C++ entry point
+
+Now, we'll implement the C++ entry point that we call from `crt1.S`. 
+
+```cpp
+{{#include ../../../sysdeps/demo/entry.cpp}}
+```
+
+The call to `__dlapi_enter` is used to perform initialisation in statically linked executables (but is a no-op in dynamically linked ones). For example, any global constructors in the program will be called from here.
+
+## Performing system calls
+
+The final piece of infrastructure we require is the ability to invoke system calls.
+
+For example, on RISC-V a system call is invoked via the `ecall` instruction and requires putting arguments in specific registers, which requires a bit of (inline) assembly. We recommend copying this glue from an existing target.
+
+For the demo OS, this is provided by `syscall.cpp` and `include/bits/syscall.h`.
+
+## Implementing sysdeps
+
+Finally we're ready to implement the actual sysdep functions. For a basic statically-linked hello world program, you'll need to provide definitions of the following sysdep functions:
+
+- `mlibc::sys_libc_panic`
+- `mlibc::sys_libc_log`
+- `mlibc::sys_isatty`
+- `mlibc::sys_write`
+- `mlibc::sys_tcb_set`
+- `mlibc::sys_anon_allocate`
+- `mlibc::sys_anon_free`
+- `mlibc::sys_seek`
+- `mlibc::sys_exit`
+- `mlibc::sys_close`
+- `mlibc::sys_futex_wake`
+- `mlibc::sys_futex_wait`
+- `mlibc::sys_read`
+- `mlibc::sys_open`
+- `mlibc::sys_vm_map`
+- `mlibc::sys_clock_get`
+
+Note that many of these functions are declared as weak symbols. You _must_ include the relevant headers (e.g `<mlibc/all-sysdeps.hpp>`) before providing definitions otherwise these won't be emitted as weak symbols and you'll see strange errors.
```

Comment:
Maybe we should better describe the breakage?
> Diverging from the declared sysdep signature will result in different mangling of the function (because they are not declared extern "C", but have C++ linkage instead), which makes it look like the sysdep is missing - this is a silent breakage that likely does not result in compiler errors or even warnings.